tls/s2n_cipher_suites.c

```diff
@@ -1168,6 +1168,108 @@ static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, u
     return 0;
 }
 
+/* Finds and sets the index of the ChaCha20 cipher in the server security policy. If ChaCha20 does not exist, then the index is set to -1. */
+static S2N_RESULT s2n_get_index_of_chacha20_in_cipher_preferences(const struct s2n_cipher_preferences* cipher_preferences, int* idx) {
+    RESULT_ENSURE_REF(cipher_preferences);
+    RESULT_ENSURE_REF(idx);
+
+    const uint8_t chacha20_iana[S2N_TLS_CIPHER_SUITE_LEN] = { TLS_CHACHA20_POLY1305_SHA256 };
+
+    for (int i = 0; i < cipher_preferences->count; i++) {
+        const uint8_t *cipher_iana = cipher_preferences->suites[i]->iana_value;
+        if (s2n_constant_time_equals(chacha20_iana, cipher_iana, S2N_TLS_CIPHER_SUITE_LEN)) {
+            *idx = i;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *idx = -1;
+    return S2N_RESULT_OK;
+}
+
+/* If the client has ChaCha20 as their most preferred cipher suite (first cipher suite) then set flag to true. */
+static S2N_RESULT s2n_wire_indicates_chacha20_boosting(uint8_t* wire, uint8_t count, uint32_t cipher_suite_len, bool* chacha20_boosted) {
```

Comment:
Can we use a stuffer or blob for the wire pointer instead? This function assumes the read is within bounds, which can be dangerous.

Reply:
What does 'assumes the read is within bounds' here mean; is the concern whether `count` is correct? If it's any better, I can try to re-purpose the existing [s2n_wire_ciphers_contain](https://github.com/aws/s2n-tls/pull/3543/files/68d68f21a7e823077be30ddd9b714c506565e34b#diff-d0b6d9d296ee7b71a99c83ad1423142d1cc7d6b62a84ebfc31a9b0f633a4f323R1158) for this use case instead of adding this new function.

Reply:
Yeah on line 1201 you're doing pointer arithmetic that can potentially go beyond the valid region of memory for `wire`. One of the biggest reasons for blob and stuffer is to avoid doing ad-hoc pointer math and make it really difficult to do the wrong thing.